src/rejectdups-rust/src/lib.rs

```diff
@@ -0,0 +1,500 @@
+) -> ProgramResult {
+
+    info!("Rust program entrypoint");
+
+    // Iterating accounts is safer then indexing
+    let accounts_iter = &mut accounts.iter();
+
+    // Get the account that holds the vote count
+    let account = next_account_info(accounts_iter)?;
+
+    // The account must be owned by the program in order to modify its data
+    if account.owner != program_id {
+        info!("Vote account does not have the correct program id");
+        return Err(ProgramError::IncorrectProgramId);
+    }
+
+    // The data must be large enough to hold two u32 vote counts
+    if account.try_data_len()? < 2 * mem::size_of::<u32>() {
+        info!("Vote account data length too small for u32");
+        return Err(ProgramError::InvalidAccountData);
+    }
+
+    // Get the account that checks for dups
+    let check_account = next_account_info(accounts_iter)?;
+
+    // The check account must be owned by the program in order to modify its data
+    if check_account.owner != program_id {
+        info!("Check account does not have the correct program id");
+        return Err(ProgramError::IncorrectProgramId);
+    }
+
+    // The data must be large enough to hold one u32 int
+    if check_account.try_data_len()? < mem::size_of::<u32>() {
+        info!("Check account data length too small for u32");
+        return Err(ProgramError::InvalidAccountData);
+    }
+
+    // The account must be rent exempt, i.e. live forever
+    let sysvar_account = next_account_info(accounts_iter)?;
+    let rent = <Rent as Sysvar>::from_account_info(sysvar_account)?;
+    if ! sysvar::rent::check_id(sysvar_account.key) {
+        info!("Rent system account is not rent system account");
+        return Err(ProgramError::InvalidAccountData);
+    }
+    if ! rent.is_exempt(check_account.lamports(),check_account.data_len()) {
+        info!("Check account is not rent exempt");
+        return Err(ProgramError::InvalidAccountData);
+    }
+
+    // the voter 
+    let voter_account = next_account_info(accounts_iter)?;
+
+    if !voter_account.is_signer {
+      return Err(ProgramError::MissingRequiredSignature);
+    }
+
+    let expected_check_account_address = create_address_with_seed(voter_account.key, 
+                                                                  "checkvote", 
+                                                                  program_id);
+
+    if expected_check_account_address != *check_account.key {
+        info!("Voter fraud! not the correct check_account.");
+        return Err(ProgramError::InvalidAccountData);
+    }
+
+    let mut checkdata = check_account.try_borrow_mut_data()?;
+
+    let cac = LittleEndian::read_u32(&checkdata[0..4]);
+
+    if cac!=0 {
+        info!("Voter fraud! You already voted.");
+        return Err(ProgramError::InvalidAccountData);
+    }
+
+
+    // Increment vote count of candidate
+
+    let mut data = account.try_borrow_mut_data()?;
+
+    if 1 == instruction_data[0] { 
```

Comment:
Implementing `solana_sdk::program_pack::Pack` for the instruction data would also be cleaner

Reply:
k